src/components/Editor/CitationManager/index.js

```diff
@@ -0,0 +1,333 @@
+   */
+  updateInfo({ title, url }) {
+    new URL(url)
+
+    if (title.length === 0) {
+      throw new CitationOpError('Citation.info: title should not be empty')
+    }
+
+    this.info = Object.assign({}, this.info, { title, url })
+
+    const ids = this.nodes.flatMap((n) => n.editorId)
+    for (const id of ids) {
+      this.nodes[0].node.$nextTick(() => {
+        const content = store.state.post.currentEditors[id].getJSON()
+        store.commit('post/UPDATE_BLOCK_CONTENT', {
+          id,
+          content
+        })
+      })
+    }
+  }
+
+  /**
+   * update index of this citation
+   * @param {Number} index new index of node to be updated
+   */
+  updateIndex(index) {
+    this.info = Object.assign({}, this.info, { index })
+  }
+
+  /**
+   * sort list of nodes from top to bottom
+   */
+  async sortNodesArray() {
+    const blocks = store.state.post.blocks.map((b) => b.id)
+
+    // create an object with key of editor ids
+    const bucket = blocks.reduce((acc, curr) => ({ ...acc, [curr]: [] }), {})
+
+    // put nodes into bucket groupped by editorId
+    for (const node of this.nodes) {
+      bucket[node.editorId].push(node)
+    }
+
+    // sort by position in editor block
+    for (const [k, v] of Object.entries(bucket)) {
+      bucket[k] = v.sort((a, b) => a.pos > b.pos)
+    }
+
+    // set sorted list in nodes
+    this.nodes = Object.values(bucket).flatMap((b) => b)
+  }
+
+  /**
+   * get last node of this citation
+   * @returns {CitationNode}
+   */
+  get lastNode() {
+    return this.nodes[this.nodes.length - 1]
+  }
+}
+
+const CitationInfoHandler = {
+  get: (target, prop) => {
+    return target[prop]
+  },
+  /**
+   * avoid setting properties directly on the target
+   * @param {CitatoinInfo} target
+   * @param {String} prop key of target
+   * @param {Object} value new value of target[key]
+   */
+  set: (target, prop, value) => {
+    if (prop !== 'info') {
+      throw new CitationOpError('Improper operation: No such property')
+    } else {
+      throw new CitationOpError(
+        'Improper operation: Should now set info directly.'
+      )
+    }
+  }
+}
+
+class CitationManager {
+  /**
+   * constructor of CitationManager
+   */
+  constructor() {
+    this.citationList = []
+    this.citationMap = new Map()
+  }
+
+  /**
+   * connect node with citation.
+   * if citation not existed, create new one.
+   * @param {String} title title of citation
+   * @param {String} url url of citation
+   * @param {Node} node node to be connected
+   * @returns {Proxy} proxy of citationInfo
+   */
+  registerNode(title, url, node) {
+    const key = `${title}-${url}`
+
+    let citationInfo = this.citationMap.get(key)
+
+    // if no citation existed, create one
+    if (!citationInfo) {
+      citationInfo = new CitationInfo(title, url)
+      this.citationList.push(citationInfo)
+      this.citationMap.set(key, citationInfo)
+    }
+
+    citationInfo.addNode(node)
```

Comment:
register node with validation

Reply:
This `node` is a vue instance of a Tiptap editor node. It will be passed into `CitationNode`. If `node` is not a vue instance of a Tiptap editor node, `CitationInfo` will throw an error because `CitationNode.pos` is `undefined`.
I suppose there's no validation required here. Every error handling should be controlled in components.